engine/consensus/compliance/engine.go

```diff
@@ -149,9 +173,108 @@ func (e *Engine) ProcessLocal(event interface{}) error {
 // Process processes the given event from the node with the given origin ID in
 // a blocking manner. It returns the potential processing error when done.
 func (e *Engine) Process(originID flow.Identifier, event interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(originID, event)
-	})
+	e.pendingEventSink <- &Event{
+		OriginID: originID,
+		Msg:      event,
+	}
+	return nil
+}
+
+// processEvents is processor of pending events which drives events from networking layer to business logic in `Core`.
+// Effectively consumes messages from networking layer and dispatches them into corresponding sinks which are connected with `Core`.
+// Should be run as a separate goroutine.
+func (e *Engine) processEvents() {
+	// takes pending event from one of the queues
+	// nil sink means nothing to send, this prevents blocking on select
+	fetchEvent := func() (*Event, EventSink, *fifoqueue.FifoQueue) {
+		if val, ok := e.pendingBlocks.Front(); ok {
+			return val.(*Event), e.blockSink, e.pendingBlocks
+		}
+		if val, ok := e.pendingVotes.Front(); ok {
+			return val.(*Event), e.voteSink, e.pendingVotes
```

Comment:
Better to include the FIFO queue in the sink, since an event sink is always backed by one queue. 

And we can hide the implementation of the event sink by wrapping the interactions to the FIFO queue with methods like:
- `sink.Next() (event Event, hasNext bool)`
- `sink.Pop()`

Reply:
I would suggest to move forward with this PR and merge it. Once we have decided on a design for the queues, lets refactor.